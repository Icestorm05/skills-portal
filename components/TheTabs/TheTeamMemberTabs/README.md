# The Team Member Tabs

## Functionality

The team member tabs component displays a list of routes available in the application. These routes include a member's: `Profile`, `Unapproved Skills`, `Approved Skills`, `Available Skills` and `Team`. This component should only be used once within the lifecycle of this application.

## Usage

### Props

`<the-team-member-tabs>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `color` | String *optional* | The color of each tab. Defaults to `secondary`. Accepts any of: `primary`, `accent`, `secondary`, `info`, `warning`, `error` or `success`.
| `hasTeam` | Boolean *optional* | Whether or not to display the team tab or not. Defaults to `false`.
| `prefix` | String | The text to display before the name of the tab item.
| `sliderColor` | String *optional* | The slider color of the selected tab. Defaults to `warning`. Accepts any of: `primary`, `accent`, `secondary`, `info`, `warning`, `error` or `success`.

### Example

```html
<the-team-member-tabs prefix="John's"/>
```

```ts
import Component from 'nuxt-class-component';
import Vue from 'vue';

import TheTeamMemberTabs from '~/components/TheTabs/TheTeamMemberTabs/TheTeamMemberTabs.vue';

@Component({
    components: {
        TheTeamMemberTabs,
    },
})
export default class GenericComponent extends Vue {}
```