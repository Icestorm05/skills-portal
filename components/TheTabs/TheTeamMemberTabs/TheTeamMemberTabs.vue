<template lang="pug">
    v-tabs(:color="color"
           grow
           icons-and-text
           show-arrows)#TheTabs
        v-tabs-slider(:color="sliderColor")
        v-tab(ripple
              nuxt
              :to="{ name: 'id-profile', params: { id: $route.params.id } }").white--text.Tab#ProfileTab {{ amendedPrefix }} Profile
              v-icon person
        v-tab(ripple
              nuxt
              exact
              :to="{ name: 'id-skills-unreviewed', params: { id: $route.params.id } }").white--text.Tab#UnreviewedSkillsTab {{ amendedPrefix }} Unreviewed Skills
              v-icon list
        v-tab(ripple
              nuxt
              exact
              :to="{ name: 'id-skills-reviewed', params: { id: $route.params.id } }").white--text.Tab#ReviewedSkillsTab {{ amendedPrefix }} Reviewed Skills
              v-icon playlist_add_check
        v-tab(ripple
              nuxt
              :to="{ name: 'id-team', params: { id: $route.params.id } }"
              v-if="hasTeam").white--text.Tab#TeamTab {{ amendedPrefix }} Team
              v-icon people
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import Vue from "vue";
import { Prop } from "vue-property-decorator";

import VIcon from "vuetify/es5/components/VIcon";
import VTabs, { VTab, VTabsSlider } from "vuetify/es5/components/VTabs";

@Component({
  components: {
    VIcon,
    VTab,
    VTabs,
    VTabsSlider,
  }
})
export default class TheTeamMemberTabs extends Vue {
  @Prop({ type: String, default: "accent" })
  private color: string;
  @Prop({ type: Boolean, default: false })
  private hasTeam: boolean;
  @Prop({ type: String, required: true })
  private prefix: string;
  @Prop({ type: String, default: "warning" })
  private sliderColor: string;

  private get amendedPrefix() {
    return this.prefix.charAt(this.prefix.length - 1) === "s"
      ? `${this.prefix}'`
      : `${this.prefix}'s`;
  }
}
</script>

<style lang="stylus" scoped>
.Tab:hover /deep/ .tabs__item {
    opacity: 1;
}
</style>
