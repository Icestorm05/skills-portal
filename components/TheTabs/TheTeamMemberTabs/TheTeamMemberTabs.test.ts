import TheTeamMemberTabs from "./TheTeamMemberTabs.vue";
import Vuetify from "vuetify/es5/components/Vuetify";
import VTabs from "vuetify/es5/components/VTabs";
import { createLocalVue, mount } from "@vue/test-utils";

declare const expect: jest.Expect;

const localVue = createLocalVue();
localVue.use(Vuetify, {
  components: {
    VTabs
  }
});

const stubs = {
  "nuxt-link": "<a></a>"
};

const mocks = {
  $route: {
    params: {
      id: 1
    }
  }
};

describe("The Tabs", () => {
  test("is a Vue instance", () => {
    const wrapper = mount(TheTeamMemberTabs, {
      localVue,
      mocks,
      stubs,
      propsData: { prefix: "John" }
    });
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("snapshot has not changed", () => {
    const wrapper = mount(TheTeamMemberTabs, {
      localVue,
      stubs,
      propsData: { prefix: "John" }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("color prop changes color of tabs", () => {
    const wrapper = mount(TheTeamMemberTabs, {
      localVue,
      mocks,
      stubs,
      propsData: { color: "error", prefix: "John" }
    });
    expect(wrapper.find("#TheTabs .tabs__bar").classes()).toContain("error");
  });

  test("slider color prop changes color of tab slider", () => {
    const wrapper = mount(TheTeamMemberTabs, {
      localVue,
      mocks,
      stubs,
      propsData: { sliderColor: "error", prefix: "John" }
    });
    expect(wrapper.find("#TheTabs .tabs__slider").classes()).toContain("error");
  });

  test("team tab is hidden by default", () => {
    const wrapper = mount(TheTeamMemberTabs, {
      localVue,
      mocks,
      stubs,
      propsData: { prefix: "John" }
    });
    expect(wrapper.find("#TheTabs #TeamTab").exists()).toBeFalsy();
  });

  test("team tab is visible with hasTeam prop", () => {
    const wrapper = mount(TheTeamMemberTabs, {
      localVue,
      mocks,
      stubs,
      propsData: { hasTeam: true, prefix: "John" }
    });
    expect(wrapper.find("#TheTabs #TeamTab").exists()).toBeTruthy();
  });

  test("prefix appends a 's to tab names", () => {
    const wrapper = mount(TheTeamMemberTabs, {
      localVue,
      mocks,
      stubs,
      propsData: { hasTeam: true, prefix: "John" }
    });
    expect(wrapper.find("#TheTabs #UnreviewedSkillsTab").text()).toContain(
      "John's Unreviewed Skills"
    );
    expect(wrapper.find("#TheTabs #ReviewedSkillsTab").text()).toContain(
      "John's Reviewed Skills"
    );
    expect(wrapper.find("#TheTabs #ProfileTab").text()).toContain(
      "John's Profile"
    );
    expect(wrapper.find("#TheTabs #TeamTab").text()).toContain("John's Team");
  });

  test("prefix appends a ' to tab names if name ends in a s", () => {
    const wrapper = mount(TheTeamMemberTabs, {
      localVue,
      mocks,
      stubs,
      propsData: { hasTeam: true, prefix: "James" }
    });
    expect(wrapper.find("#TheTabs #UnreviewedSkillsTab").text()).toContain(
      "James' Unreviewed Skills"
    );
    expect(wrapper.find("#TheTabs #ReviewedSkillsTab").text()).toContain(
      "James' Reviewed Skills"
    );
    expect(wrapper.find("#TheTabs #ProfileTab").text()).toContain(
      "James' Profile"
    );
    expect(wrapper.find("#TheTabs #TeamTab").text()).toContain("James' Team");
  });
});
