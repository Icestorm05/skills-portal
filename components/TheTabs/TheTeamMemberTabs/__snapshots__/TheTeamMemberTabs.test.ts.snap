// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Tabs snapshot has not changed 1`] = `
<div class="tabs" id="TheTabs">
    <div class="tabs__bar accent">
        <!---->
        <div class="tabs__wrapper">
            <div class="tabs__container tabs__container--grow tabs__container--icons-and-text">
                <div class="tabs__slider-wrapper">
                    <div class="tabs__slider warning"></div>
                </div>
                <div class="tabs__div white--text Tab" id="ProfileTab">
                    <nuxt-link class="tabs__item">John's Profile
                        <i aria-hidden="true" class="icon material-icons">person</i>
                    </nuxt-link>
                </div>
                <div class="tabs__div white--text Tab" id="UnreviewedSkillsTab">
                    <nuxt-link class="tabs__item">John's Unreviewed Skills
                        <i aria-hidden="true" class="icon material-icons">list</i>
                    </nuxt-link>
                </div>
                <div class="tabs__div white--text Tab" id="ReviewedSkillsTab">
                    <nuxt-link class="tabs__item">John's Reviewed Skills
                        <i aria-hidden="true" class="icon material-icons">playlist_add_check</i>
                    </nuxt-link>
                </div>
                <!---->
            </div>
        </div>
        <!---->
    </div>
</div>
`;
